File: analytics/src/main/java/io/appmetrica/analytics/impl/ReportsSender.java
package io.appmetrica.analytics.impl;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor;
import io.appmetrica.analytics.impl.client.ProcessConfiguration;
import io.appmetrica.analytics.impl.service.ServiceDataReporter;
import io.appmetrica.analytics.impl.service.commands.ReportToServiceCallable;
import io.appmetrica.analytics.impl.service.commands.ServiceCallableFactory;
import io.appmetrica.analytics.logger.internal.YLogger;
import java.util.concurrent.TimeUnit;

public class ReportsSender implements ServiceDataReporter {

    private static final String TAG = "[ReportsSender]";

    private final AppMetricaConnector mServiceConnector;
    private final ServiceCallableFactory serviceCallableFactory;

    private final ICommonExecutor mExecutor;

    public ReportsSender(
        @NonNull final AppMetricaConnector appMetricaConnector,
        @NonNull final ServiceCallableFactory serviceCallableFactory
    ) {
        this(
            appMetricaConnector,
            serviceCallableFactory,
            ClientServiceLocator.getInstance().getClientExecutorProvider().getReportSenderExecutor()
        );
    }

    @VisibleForTesting
    ReportsSender(
        @NonNull final AppMetricaConnector appMetricaConnector,
        @NonNull final ServiceCallableFactory serviceCallableFactory,
        @NonNull final ICommonExecutor executor
    ) {
        mExecutor = executor;
        this.serviceCallableFactory = serviceCallableFactory;
        mServiceConnector = appMetricaConnector;
    }

    public void queueReport(ReportToSend report) {
        mExecutor.submit(report.isCrashReport() ?
            serviceCallableFactory.createCrashCallable(report) :
            serviceCallableFactory.createReportCallable(report)
        );
    }

    public void sendCrash(@NonNull ReportToSend report) {
        ReportToServiceCallable callable = serviceCallableFactory.createCrashCallable(report);
        if (mServiceConnector.isConnected()) {
            try {
                mExecutor.submit(callable).get(4, TimeUnit.SECONDS);
            } catch (Throwable e) {
                YLogger.error(TAG, e);
            }
        }
        // flag will be false if it is still executing but check inside CrashCallable will prevent duplicates
        if (!callable.isExecuted()) {
            try {
                callable.call();
            } catch (Throwable ex) {
                YLogger.error(TAG, ex);
            }
        }
    }

    @Override
    public void reportData(int type, @NonNull Bundle bundle) {
        mExecutor.submit(serviceCallableFactory.createTypedReportCallable(type, bundle));
    }

    public void queueResumeUserSession(@NonNull final ProcessConfiguration processConfiguration) {
        mExecutor.submit(serviceCallableFactory.createResumeUseSessionCallable(processConfiguration));
    }

    public void queuePauseUserSession(@NonNull final ProcessConfiguration processConfiguration) {
        mExecutor.submit(serviceCallableFactory.createPauseUseSessionCallable(processConfiguration));
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/ReportsSender.kt
package io.appmetrica.analytics.impl

import android.os.Bundle
import androidx.annotation.VisibleForTesting
import io.appmetrica.analytics.coreutils.internal.time.SystemTimeProvider
import io.appmetrica.analytics.coreutils.internal.time.TimeProvider
import io.appmetrica.analytics.impl.client.ProcessConfiguration
import io.appmetrica.analytics.impl.service.ServiceDataReporter
import io.appmetrica.analytics.impl.service.commands.ServiceCallableFactory
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger
import java.util.concurrent.TimeUnit
import kotlin.math.max

class ReportsSender @VisibleForTesting internal constructor(
    private val serviceConnector: AppMetricaConnector,
    private val serviceCallableFactory: ServiceCallableFactory,
    private val timeProvider: TimeProvider
) : ServiceDataReporter {
    private val executor = ClientServiceLocator.getInstance().clientExecutorProvider.reportSenderExecutor

    constructor(
        appMetricaConnector: AppMetricaConnector,
        serviceCallableFactory: ServiceCallableFactory
    ) : this(
        appMetricaConnector,
        serviceCallableFactory,
        SystemTimeProvider()
    )

    fun queueReport(report: ReportToSend) {
        executor.submit(
            if (report.isCrashReport) {
                serviceCallableFactory.createCrashCallable(report)
            } else {
                serviceCallableFactory.createReportCallable(report)
            }
        )
    }

    fun sendCrash(report: ReportToSend) {
        val crashProcessingStartTime = timeProvider.uptimeMillis()
        val callable = serviceCallableFactory.createCrashCallable(report)
        if (serviceConnector.isConnected) {
            try {
                DebugLogger.info(TAG, "Execute crash callable asynchronous and wait")
                executor.submit(callable).get(CRASH_PROCESSING_TIME_LIMIT, TimeUnit.MILLISECONDS)
            } catch (e: Throwable) {
                DebugLogger.error(TAG, e)
            }
        }
        // flag will be false if it is still executing but check inside CrashCallable will prevent duplicates
        if (!callable.isExecuted) {
            try {
                DebugLogger.info(TAG, "Execute crash callable synchronous")
                callable.call()
            } catch (ex: Throwable) {
                DebugLogger.error(TAG, ex)
            }
        } else {
            DebugLogger.info(TAG, "Crash callable is already executed. Ignore it")
        }
        val processingTimeSpent = timeProvider.uptimeMillis() - crashProcessingStartTime
        val remainingTime = max(0, CRASH_PROCESSING_TIME_LIMIT - processingTimeSpent)
        waitFor(remainingTime)
    }

    private fun waitFor(timeout: Long) {
        try {
            DebugLogger.error(TAG, "Wait for $timeout ms to process all events before crash")
            Thread.sleep(timeout)
        } catch (e: Throwable) {
            DebugLogger.error(TAG, e)
        }
    }

    override fun reportData(type: Int, bundle: Bundle) {
        executor.submit(serviceCallableFactory.createTypedReportCallable(type, bundle))
    }

    fun queueResumeUserSession(processConfiguration: ProcessConfiguration) {
        executor.submit(serviceCallableFactory.createResumeUseSessionCallable(processConfiguration))
    }

    fun queuePauseUserSession(processConfiguration: ProcessConfiguration) {
        executor.submit(serviceCallableFactory.createPauseUseSessionCallable(processConfiguration))
    }

    companion object {
        private const val TAG = "[ReportsSender]"

        private val CRASH_PROCESSING_TIME_LIMIT = TimeUnit.SECONDS.toMillis(4)
    }
}
